WalletWasabi/Crypto/Api/CredentialIssuer.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Crypto.ZeroKnowledge.LinearRelation;
+using WalletWasabi.Crypto.ZeroKnowledge.NonInteractive;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.Api
+{
+	public class CredentialIssuer
+	{
+		// Keeps track of the used serial numbers. This is part of 
+		// the double-spending prevention mechanism.
+		private HashSet<GroupElement> SerialNumbers { get; } = new HashSet<GroupElement>();
+		
+		// Canary test check to ensure credential balance is never negative
+		private Money Balance { get; set; } = Money.Zero;
+
+		private WasabiRandom Random { get; }
+
+		private CoordinatorSecretKey CoordinatorSecretKey { get; }
+		
+		private CoordinatorParameters CoordinatorParameters { get; }
+
+		public int NumberOfCredentials { get; }
+
+		public CredentialIssuer(CoordinatorSecretKey sk, int numberOfCredentials, WasabiRandom random)
+		{
+			CoordinatorSecretKey = Guard.NotNull(nameof(sk), sk);
+			NumberOfCredentials = Guard.InRangeAndNotNull(nameof(numberOfCredentials), numberOfCredentials, 1, 100);
+			CoordinatorParameters = CoordinatorSecretKey.ComputeCoordinatorParameters();
+			Random = random;
+		}
+
+		public RegistrationResponse HandleRequest(RegistrationRequest registrationRequest)
+		{
+			Guard.NotNull(nameof(registrationRequest), registrationRequest);
+			Guard.Same(nameof(registrationRequest), NumberOfCredentials, registrationRequest.Requested.Count());
+			Guard.Same(nameof(registrationRequest), registrationRequest.IsNullRequest ? 0 : NumberOfCredentials, registrationRequest.Presented.Count());
+
+			// Don't allow balance to go negative. In case this goes below zero 
+			// then there is a problem somewhere because this should not be possible.
+			if (Balance + registrationRequest.Balance < Money.Zero)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.NegativeBalance);
+			}
+
+			// Check all the bit commitments have the correct length. Null requests need zero-length rangeproofs 
+			var rangeProofWidth = registrationRequest.IsNullRequest ? 0 : Constants.RangeProofWidth;
+			var allRangeProofsAreCorrentSize = registrationRequest.Requested.All(x => x.BitCommitments.Count() == rangeProofWidth);
+			if (!allRangeProofsAreCorrentSize)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.InvalidBitCommitment);
+			} 
+
+			var sn = SerialNumbers.ToArray(); 
```

Comment:
Isn't this check superfluous, because this is anyways caught after a few lines, when coordinator iterates over the `registrationRequest.Presented`? If this is deemed to be an overkill, then there is enough to have a single Exception for repeated serial numbers.  So e.g. `WabiSabiErrorCode.SerialNumberDuplicated` could be omitted.

Reply:
No, it is not superflous.